### PR TITLE
feat: `Name` trait + `Any` encoding support

### PR DESCRIPTION
**Pull request by tarcieri**
As discussed in #299 and #858, adds a `Name` trait which associates a type name and package constants with a `Message` type. It also provides `full_name` and `type_url` methods.

The `type_url` method is used by newly added methods on the `Any` type which can be used for decoding/encoding messages:

- `Any::from_msg`: encodes a given `Message`, returning `Any`.
- `Any::to_msg`: decodes `Any::value` as the given `Message`, first validating the message type has the expected type URL.

### CODE REVIEW

File: prost-types/src/lib.rs
Comment (tarcieri) [inline prost-types/src/lib.rs:60]: This check is naive. Apparently we should split at "/" and compare the fullnames afterward to see if they match

**Comment by tarcieri** [comment]
@LucioFranco I'm trying to interop with a Go reference implementation which makes pretty extensive use of `Any`, though that said, I am also trying to figure out which things are standard conventions and which are idioms of my particular implementation, as we observed in #858.

That said you did make me take a second look and I did notice something a bit problematic in the type URL comparison. I'm not sure how other implementations handle the leading domain component of a type URL, i.e. if it mismatches, should the full type URL mismatch. Should type URLs with leading `/` be treated as "relative"?

All that said, I deliberately made `type_url` a provided method so it implements a convention but can be overridden if desired.

We might also consider an associated constant for the domain component, which could be blank.

File: src/name.rs
Comment (LucioFranco) [inline src/name.rs:7]: Can this just live in `prost_types`, why does it need to live in the core crate?

**Comment by LucioFranco** [comment]
@tarcieri this looks good besides my one question, I am okay with merging mostly because a lot of this code can be kept outside the core crate.

**Comment by vitruvvius** [comment]
nice !!!

File: src/name.rs
Comment (tarcieri) [reply src/name.rs:7]: I was hoping it could eventually be autogenerated by `prost-build`, at least optionally. We're hand annotating them right now and it's getting quite cumbersome.

It seems like `prost-build` currently only refers to types/traits in `prost`. If it were optional and it could refer to types/traits in `prost-types`, I'd be fine with the `Name` trait living there.

File: prost-types/src/lib.rs
Comment (tarcieri) [reply prost-types/src/lib.rs]: I added more robust type URL matching in [01fde2d](https://github.com/tokio-rs/prost/pull/896/commits/01fde2daebb7e75d5376a560ee50ac900119890c)

**Comment by tarcieri** [comment]
Build failure seems unrelated:

https://github.com/tokio-rs/prost/actions/runs/5981227819/job/16228748609?pr=896#step:7:112

```
error: package `petgraph v0.6.4` cannot be built because it requires rustc 1.64 or newer, while the currently active rustc version is 1.63.0
Error: Process completed with exit code 101.
```

**Comment by tarcieri** [comment]
Regardless of what crate `Name` ends up living in, I think it'd also be good to impl this trait for the `Message` types in `prost-types`.

That would also make it possible to test `Any::{from_msg, to_msg}`.

**Comment by tarcieri** [comment]
Aside from the question of where the `Name` trait should live, this should be ready to merge

File: src/name.rs
Comment (tarcieri) [inline src/name.rs:24]: Perhaps this could include some of the documentation I added to the internal `TypeUrl` type.

File: src/name.rs
Comment (tarcieri) [inline src/name.rs:26]: This could potentially benefit from an easier way to customize the URL authority (i.e. hostname+port). It was a bit tricky to do this for the "well known" types:

https://github.com/tokio-rs/prost/pull/896/files#diff-fe4eba36fffb306e25d27bf29ddea9cd1634ea4ad86705045c1cb13a3aa27346R536-R540

File: src/name.rs
Comment (LucioFranco) [reply src/name.rs:7]: I see your point, do we just want to go ahead and do the work to also add the prost-build parts? The next release will be breaking and I would be ok if we add something like that.

File: src/name.rs
Comment (LucioFranco) [inline src/name.rs:18]: Should these be `&'static str`, is there a case where these are not statically defined with the generated proto?

**Comment by LucioFranco** [comment]
@tarcieri sorry for dragging this a long, would you be interested in adding the prost-build part too then we can keep the trait in prost. I merged the msrv fix in another PR so would be good to back that out of this one. We could also merge this part now (once the msrv is backed out and this is rebased) and then we can add the prost-build stuff in a follow up.

File: src/name.rs
Comment (tarcieri) [reply src/name.rs:18]: If `prost-build` filled them in they potentially could be all computed at build-time.

Otherwise it's not really possible to construct a `&'static str` as a concatenation of `Self::PACKAGE` + `'.'` + `Self::NAME` AFAIK.

**Comment by tarcieri** [comment]
I can attempt to add the `prost-build` part, although I'm not sure how difficult that will be. Hopefully it has all of the needed information a priori

**Comment by LucioFranco** [comment]
@tarcieri ok cool, let me know if I can help (feel free to ping on discord). I think I want to wait on this for the next (breaking) release which will basically happen after we get this merged.

**Comment by LucioFranco** [comment]
I'd like to get things merged and released today so I will go ahead and get this merged and figure out what else we need today and then we can add more stuff in point releases since they shouldn't be breaking changes.

**Review by LucioFranco — APPROVED** [review]

**Comment by sillykelvin** [comment]
Sorry to disturb, is there any progress on the `prost-build` part? Currently we are writing `impl prost::Name for Foo { ... }` manually, however it's awful and error-prone, it would be great if we can do this with `prost-build`.

**Comment by LucioFranco** [comment]
@tarcieri were you able to follow up on this? If not if anyone else is interested in picking this up?

**Comment by tarcieri** [comment]
Not yet, and I'm on vacation. Would be great if someone else could pick it up.

**Comment by LucioFranco** [comment]
@tarcieri enjoy! Sounds good!

Well if someone wants to pick up this work I opened https://github.com/tokio-rs/prost/issues/926 and we can continue discussing there.

**Review by Missjones2 — APPROVED** [review]

**Review by Missjones2 — APPROVED** [review]

**Pull request merged**
